crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,529 @@
+        broadcast_sender: broadcast::Sender<NativePriceResult>,
+    ) -> JoinHandle<()> {
+        tokio::task::spawn(batched_for_each(config, calls, move |batch| {
+            let inner = inner.clone();
+            let broadcast_sender = broadcast_sender.clone();
+            async move {
+                let batch = batch.into_iter().collect::<HashSet<_>>();
+                if !batch.is_empty() {
+                    let results: Vec<_> = match inner.fetch_native_prices(&batch).await {
+                        Ok(results) => results
+                            .into_iter()
+                            .map(|(token, price)| NativePriceResult {
+                                token,
+                                result: price,
+                            })
+                            .collect(),
+                        Err(err) => {
+                            tracing::error!(?err, "failed to send native price batch request");
+                            batch
+                                .into_iter()
+                                .map(|token| NativePriceResult {
+                                    token,
+                                    result: Err(err.clone()),
+                                })
+                                .collect()
+                        }
+                    };
+                    for result in results {
+                        let _ = broadcast_sender.send(result);
+                    }
+                }
+            }
+        }))
+    }
+
+    /// Request to get estimate prices in a batch
+    pub async fn request_buffered_estimate_prices(
+        &self,
+        token: &H160,
+    ) -> NativePriceEstimateResult {
+        // Sends the token for requesting price
+        self.calls.unbounded_send(*token).map_err(|e| {
+            PriceEstimationError::ProtocolInternal(anyhow!(
+                "failed to append a new token to the queue: {e:?}"
+            ))
+        })?;
+
+        let mut rx = self.broadcast_sender.subscribe();
+
+        tokio::time::timeout(self.config.result_ready_timeout, async {
+            loop {
+                if let Ok(Some(result)) =
+                    Self::receive_with_timeout(&mut rx, token, self.config.batch_delay).await
+                {
+                    return result.result;
+                }
+            }
+        })
+        .await
+        .map_err(|_| {
+            PriceEstimationError::ProtocolInternal(anyhow!(
+                "blocking buffered estimate prices timeout elapsed"
+            ))
+        })?
+    }
+
+    // Function to receive with a timeout
+    async fn receive_with_timeout(
+        rx: &mut broadcast::Receiver<NativePriceResult>,
+        token: &H160,
+        timeout_duration: Duration,
+    ) -> Result<Option<NativePriceResult>, Elapsed> {
+        tokio::time::timeout(timeout_duration, async {
+            match rx.recv().await {
+                Ok(value) => (value.token == *token).then_some(value),
```

Comment:
I believe this is the part that trips be up about the broadcast channel.
Presumably there are multiple listeners on the same channel interested in different responses.
I think what makes more sense to me is if there is a new oneshot channel for every caller that tries to request a token price. Then the batching mechanism would aggregate a map `Token => Vec<oneshot_channel>` and when the background task gets the prices in a batch it knows exactly who is actually awaiting this price.

Reply:
@MartinquaXD the timeout cannot be avoided even for the solution you propose. We cannot be waiting forever for the answer and block the petition. The timeout must be applied in any case.

The solution you propose it way more complex, error/bug/memory leak prone. The way of doing it would be to send in the request `mspc` the token alongside the channel to which the thread should send the token. But I don't see how that solution is better than the broadcast :thinking:

Reply:
Allocating oneshot channel per native price request is quite memory wasteful. We need to keep in mind that for the orderbook we are getting quite a lot of native price requests.

Reply:
My main concern with the broadcast is the processing time within the `match` statement, if it is too slow, we may miss events, but since it is just an `if`, it is impossible.

Reply:
I totally see your point and there are cons and pros for each solution, to be honest. I don't see a clear "winning" solution. I am more incline for the broadcast because it is way less resources intensive than the oneshot solution.

Reply:
Not sure what the timeouts have to do with my comment. Overall I agree that having timeouts would be good (although the double timeout this PR currently has looks suspicious).

So one concern I see with the broadcast channel is that you have to decide on a size upfront which has to be big enough to not cause issues even when we get spammed. That means it likely needs to be vastly overallocated for 99% of the time which kind of makes the oneshot channel approach more appealing in terms of memory usage IMO.
Also AFAIK every task touching this batching implementation might be woken up for every incoming price of the batch even if it's just interested in a single token.
And lastly as to the point of being error prone: oneshot channels are very simple while broadcast channels have kind of subtle semantics around them. For example where you currently completely discard a price whenever the channel falls behind there is actually no reason to assume that there is no price for the requested token left in the channel. (i.e. [here](https://github.com/cowprotocol/services/pull/2869/files#diff-2e490a59194917c3d8edfe352221e75a0e0a05312c50f11ec27b635cf3142334R188) you could `try_recv` the stream to completion hoping to get lucky that the token price is still in the stream).

Reply:
@MartinquaXD that is a very valid point. I myself don't know what's the best answer for this. Let's try to figure it out together.

> So one concern I see with the broadcast channel is that you have to decide on a size upfront which has to be big enough to not cause issues even when we get spammed. 

We cannot get **that** spammed anymore, but I agree. I believe the broadcast items have a very short time to live. So in theory they won't be in the channel for much time. I set it to 200 elements which don't seem to me like a huge number in terms of memory.

> That means it likely needs to be vastly overallocated for 99% of the time which kind of makes the oneshot channel approach more appealing in terms of memory usage IMO.

Umm, I am not so sure about it. The batch size for CoinGecko is 20, let's say we get 1000 request. so we cannot keep more than 20 elements in the broadcast (because the request time to CoinGecko is greater than the time to live of the items), but we definitely can keep more than 20 oneshots, because we keeps gathering and sending the requesters. And oneshop entity/object is more memory intensive than a pair `(H160, price)`.

> Also AFAIK every task touching this batching implementation might be woken up for every incoming price of the batch even if it's just interested in a single token.

This is totally true. This is a big con of the broadcast.

> And lastly as to the point of being error prone: oneshot channels are very simple while broadcast channels have kind of subtle semantics around them. For example where you currently completely discard a price whenever the channel falls behind there is actually no reason to assume that there is no price for the requested token left in the channel.

Yes, 100 % true, the issue with broadcast is that it could happen that sometimes we miss the price. This could happen in the loop in which we fire the broadcast with all the results. It could happen that the consumer miss a result while checking if the previous one was for its token.

Reply:
So far only the memory leak argument is very convincing to me. Overall I think the `RequestSharing` struct would be a perfect fit here but unfortunately this is already involved in a memory leak investigation.
If that would not have been the case I would suggest a setup where the buffering struct holds a `RequestSharing` internally which maps `Token => oneshot_channel`. Then the background task would always have to notify exactly a single channel since the requests already get deduplicated by the `RequestSharing` struct.
But until we figure out how we still leak memory in the API I would hesitant to use that component even more. 🤔

Reply:
> The batch size for CoinGecko is 20, let's say we get 1000 request. so we cannot keep more than 20 elements in the broadcast (because the request time to CoinGecko is greater than the time to live of the items), but we definitely can keep more than 20 oneshots, because we keeps gathering and sending the requesters.

The thing I don't like about the broadcast channel is that we can't make it bulletproof. In your example it's definitely not enough to have a channel size of 20 because the behavior of the solution very much depends on how quickly the consumers read the channel. If there is 0 scheduling overhead and tasks would get woken up immediately 20 slots would be fine. But let's say the runtime let's the background task run which produces more prices but for some reason doesn't schedule the consumer tasks to run. In that case the consumers might not get any prices although the background task successfully fetched all of them.
Overall the broadcast channels feel like a heuristic while oneshot channels could give us hard guarantees. And with the idea discussed above we would have extremely cheap deduplication of requests (basically the cost of an `Arc` per duplicated request).

But again given that the memory leak involving the `RequestSharing` is still unclear it indeed seems more pragmatic to go with the broadcast solution for now. Also one thing that should indicate that a relatively small buffer size is okay is that a single coingecko request is orders of magnitude slower than waking up a task for a new price meaning the producer is way slower than any of the consumers.

Reply:
Yeah, I fully agree with you! one thing

> [..] internally which maps Token => oneshot_channel

This was another reason to do the broadcast. We cannot do that mapping, we would need to do a `request_id` (we can generate a new `UUID` for it`) => oneshot_channel. Because of this scenario:

- User number 1 request price for DOGE
- User number 2 request price for DOGE at the same time (difference of ~100 ms)
- Request sent to the buffered CoinGecko with the same token for two different receivers, one of them will never receive the price as it would be squashed in the map.

I know this is a **very** remote case. But it can definitely happen too.

Reply:
> one of them will never receive the price as it would be squashed in the map.

Which is why I suggested a map like `Token => Vec<oneshot_channel>`. But doesn't matter anymore anyway.

Reply:
> The thing I don't like about the broadcast channel is that we can't make it bulletproof.

100 % true. I couldn't agree more.

> In your example it's definitely not enough to have a channel size of 20 because the behavior of the solution very much depends on how quickly the consumers read the channel.

Yes, but the client's (token requester) in theory, should be faster. Because the time to get a new bulk of 20 is the request to CoinGecko's server + [this PR](https://github.com/cowprotocol/services/pull/2876) processing the response). The logic in our side is just a loop and a token comparison. If our CPU is overloaded and we are slow processing, we would also be slow processing the CoinGecko's response prior to sending it to the broadcast. But I fully agree, that the scenario may be unlikely, but as you said, it is not bulletproof and it can definitely happen!

Either way, we should closely follow the broadcast performance and see how many prices we really miss.

Reply:
> Which is why I suggested a map like Token => Vec<oneshot_channel>. But doesn't matter anymore anyway.

Oh yeah, missed that detail! you are right!! :ok_hand: